test_libs/pyspec/eth2spec/test/block_processing/test_process_early_derived_secret_reveal.py

```diff
@@ -0,0 +1,130 @@
+import pytest
+
+from eth2spec.test.helpers.custody import get_valid_early_derived_secret_reveal
+from eth2spec.test.context import with_phase1, spec_state_test, expect_assertion_error
+
+
+def run_early_derived_secret_reveal_processing(spec, state, randao_key_reveal, valid=True):
+    """
+    Run ``process_randao_key_reveal``, yielding:
+      - pre-state ('pre')
+      - randao_key_reveal ('randao_key_reveal')
+      - post-state ('post').
+    If ``valid == False``, run expecting ``AssertionError``
+    """
+    yield 'pre', state
+    yield 'randao_key_reveal', randao_key_reveal
+
+    if not valid:
+        expect_assertion_error(lambda: spec.process_early_derived_secret_reveal(state, randao_key_reveal))
+        yield 'post', None
+        return
+
+    spec.process_early_derived_secret_reveal(state, randao_key_reveal)
+
+    slashed_validator = state.validator_registry[randao_key_reveal.revealed_index]
+
+    if randao_key_reveal.epoch >= spec.get_current_epoch(state) + spec.CUSTODY_PERIOD_TO_RANDAO_PADDING:
+        assert slashed_validator.slashed
+        assert slashed_validator.exit_epoch < spec.FAR_FUTURE_EPOCH
+        assert slashed_validator.withdrawable_epoch < spec.FAR_FUTURE_EPOCH
+    # lost whistleblower reward
+    # FIXME: Currently broken because get_base_reward in genesis epoch is 0
+    # assert (
+    #     state.balances[randao_key_reveal.revealed_index] <
+    #     state.balances[randao_key_reveal.revealed_index]
+    # )
+    yield 'post', state
+
+
+@with_phase1
+@spec_state_test
+def test_success(spec, state):
+    randao_key_reveal = get_valid_early_derived_secret_reveal(spec, state)
+
+    yield from run_early_derived_secret_reveal_processing(spec, state, randao_key_reveal)
+
+
+@with_phase1
+@spec_state_test
+def test_reveal_from_current_epoch(spec, state):
+    randao_key_reveal = get_valid_early_derived_secret_reveal(spec, state, spec.get_current_epoch(state))
+
+    yield from run_early_derived_secret_reveal_processing(spec, state, randao_key_reveal, False)
+
+
+@with_phase1
+@spec_state_test
+def test_reveal_from_past_epoch(spec, state):
+    if spec.get_current_epoch(state) < 1:
```

Comment:
Does this ever run? I'm pretty sure we are only passing in "genesis" states to tests currently.
Could just run `next_epoch` a few times to get past the start epoch

Reply:
```
    next_epoch(spec, state)
    apply_empty_block(spec, state)
```